src/Validator.ts

```diff
@@ -0,0 +1,495 @@
+  private results: ValidationResults;
+
+  constructor(schema: SchemaDefinition, options: ValidatorOptions = {}) {
+    this.schema = schema;
+    this.valueValidatorMap = new Map();
+    this.dateFormat = options.dateFormat || "yyyy-MM-dd";
+    this.dateTimeFormat = options.dateTimeFormat || "yyyy-MM-dd HH:mm";
+    this.identifiers = [];
+    this.results = {};
+  }
+
+  useTargetClass(className: string) {
+    const classDefinition = this.schema.classes?.[className];
+    if (classDefinition === undefined) {
+      throw new Error(`No class named '${className}'`);
+    }
+
+    this.results = {};
+    this.targetClass = classDefinition;
+    this.targetClassInducedSlots = classDefinition.attributes;
+
+    this.identifiers = Object.values(this.targetClassInducedSlots!)
+      .filter((slot) => slot.identifier !== undefined && slot.identifier)
+      .map((slot) => slot.name);
+
+    this.valueValidatorMap = new Map();
+  }
+
+  getValidatorForSlot(
+    slot: AnonymousSlotExpression | SlotDefinition | string,
+    options: GetValidatorsForSlotOptions = {},
+  ): ValidatorFn {
+    const { cacheKey, inheritedRange } = options;
+    if (typeof cacheKey === "string" && this.valueValidatorMap.has(cacheKey)) {
+      return this.valueValidatorMap.get(cacheKey)!;
+    }
+
+    let slotDefinition: AnonymousSlotExpression | SlotDefinition;
+    if (typeof slot === "string") {
+      slotDefinition = this.targetClassInducedSlots![slot];
+    } else {
+      slotDefinition = slot;
+    }
+
+    if (!slotDefinition.range && inheritedRange) {
+      slotDefinition.range = inheritedRange;
+    }
+
+    const slotType = this.schema.types?.[slotDefinition.range!];
+    const slotEnum = this.schema.enums?.[slotDefinition.range!];
+    const slotPermissibleValues = Object.values(
+      slotEnum?.permissible_values ?? {},
+    ).map((pv) => pv.text);
+
+    const anyOfValidators = (slotDefinition.any_of ?? []).map((subSlot) =>
+      this.getValidatorForSlot(subSlot, {
+        inheritedRange: slotDefinition.range,
+      }),
+    );
+    const allOfValidators = (slotDefinition.all_of ?? []).map((subSlot) =>
+      this.getValidatorForSlot(subSlot, {
+        inheritedRange: slotDefinition.range,
+      }),
+    );
+    const exactlyOneOfValidators = (slotDefinition.exactly_one_of ?? []).map(
+      (subSlot) =>
+        this.getValidatorForSlot(subSlot, {
+          inheritedRange: slotDefinition.range,
+        }),
+    );
+    const noneOfValidators = (slotDefinition.none_of ?? []).map((subSlot) =>
+      this.getValidatorForSlot(subSlot, {
+        inheritedRange: slotDefinition.range,
+      }),
+    );
+
+    const slotMinimumValue = this.parseMinMaxConstraint(
+      slotDefinition.minimum_value,
+      slotType?.uri,
+    );
+    const slotMaximumValue = this.parseMinMaxConstraint(
+      slotDefinition.maximum_value,
+      slotType?.uri,
+    );
+
+    const validate = (value: unknown) => {
+      const valueDefined = value != null && value !== "";
```

Comment:
Maybe this is a bad habit of mine but I tend to use `!=` or `==` when comparing with `null` because it implicitly also compares against `undefined`. So this:

```js
value != null
```

is equivalent to

```js
value !== null && value !== undefined
```

I feel somewhat justified in using that shortcut because ESLint carves out exceptions to their `eqeqeq` rule to allow such things: https://eslint.org/docs/latest/rules/eqeqeq.

Bigger picture, there are probably a lot places where we could tighten up our typing (e.g. to only allow undefined and not null), but I'm willing to save that for another time.